lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs

```diff
@@ -104,19 +113,53 @@ public byte[] Call(byte[] memento)
                 }
             }
 
-            MapInputWithControlMessage<TMapInput> stopMessage =
+            if (!_break)
+            {
+                MapInputWithControlMessage<TMapInput> stopMessage =
                     new MapInputWithControlMessage<TMapInput>(MapControlMessage.Stop);
-            _dataAndControlMessageSender.Send(stopMessage);
+                _dataAndControlMessageSender.Send(stopMessage);
+            }
 
             _resultHandler.Dispose();
+            _stopped = true;
             return null;
         }
 
         /// <summary>
+        /// Received close event. If it is closed command sent from driver, set _break to true to stop the iteration in Call method
+        /// After _enforceCloseTimeout, if the task is still not closed, throw IMRUTaskSystemException to enforce stop. 
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            if (closeEvent.Value.IsPresent() && Encoding.UTF8.GetString(closeEvent.Value.Value).Equals(TaskManager.CloseTaskByDriver))
+            {
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```

Comment:
After close handler set _break to true, it would expect the task to return by itself. The code flow is not always back to the close handler. 
Close handler wait and then enforce to close is a backup plan, it is for the case if the task hung and cannot return from Call() by itself. In this case, we must enforce to close. 
I am not sure how Semaphores can be used in this scenario. Can you elaborate?